File: cpp/july/day_9_Maximum Width of Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        int i;
        unsigned long left = 0, right = 0, s, ans = 0;
        queue<pair<TreeNode*, unsigned long>> q;
        q.push({root, 0});
        while(!q.empty()){
            s = q.size();
            left = right = 0;
            for(i = 0; i < s; i++){
                auto t = q.front();
                q.pop();
                if(i == 0){
                    left = t.second;
                }
                if(i == s - 1){
                    right = t.second;
                }
                if(t.first->left){
                    q.push({t.first->left, 2 * t.second});
                }
                if(t.first->right){
                    q.push({t.first->right, 2 * t.second + 1});
                }
            }
            ans = max(ans, right - left + 1);
        }
        return ans;
    }
};

File: python3/july/day_9_Maximum Width of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: TreeNode) -> int:
        ans = 0
        q = collections.deque([(root, 0)])
        while(len(q)):
            s = len(q)
            left = right = 0
            for i in range(s):
                t = q.popleft()
                if(i == 0):
                    left = t[1]

                if(i == s - 1):
                    right = t[1]

                if(t[0].left):
                    q.append((t[0].left, 2 * t[1]))

                if(t[0].right):
                    q.append((t[0].right, 2 * t[1] + 1))
                
            
            ans = max(ans, right - left + 1)

        return ans
